x/qgb/abci_test.go

```diff
@@ -3,75 +3,137 @@ package qgb_test
-	require.True(t, found)
 	require.Nil(t, err)
+	require.True(t, found)
 	require.NotNil(t, attestation)
 	require.Equal(t, uint64(1), attestation.GetNonce())
+
+	// get the valset
+	require.Equal(t, types.ValsetRequestType, attestation.Type())
+	vs, ok := attestation.(*types.Valset)
+	require.True(t, ok)
+	require.NotNil(t, vs)
 }
 
-func TestValsetCreationUponUnbonding(t *testing.T) {
+func TestValsetCreationWhenValidatorUnbonds(t *testing.T) {
 	input, ctx := testutil.SetupFiveValChain(t)
 	pk := input.QgbKeeper
 
-	currentValsetNonce := pk.GetLatestAttestationNonce(ctx)
-	vs, err := pk.GetCurrentValset(ctx)
-	require.Nil(t, err)
-	err = pk.SetAttestationRequest(ctx, &vs)
-	require.Nil(t, err)
+	// run abci methods after chain init
+	staking.EndBlocker(input.Context, input.StakingKeeper)
+	qgb.EndBlocker(ctx, *pk)
+
+	// current attestation expectedNonce should be 1 because a valset has been emitted upon chain init.
+	currentAttestationNonce := pk.GetLatestAttestationNonce(ctx)
+	require.Equal(t, uint64(1), currentAttestationNonce)
 
 	input.Context = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
-	// begin unbonding
 	msgServer := stakingkeeper.NewMsgServerImpl(input.StakingKeeper)
+
 	undelegateMsg := testutil.NewTestMsgUnDelegateValidator(testutil.ValAddrs[0], testutil.StakingAmount)
-	_, err = msgServer.Undelegate(input.Context, undelegateMsg)
+	_, err := msgServer.Undelegate(input.Context, undelegateMsg)
 	require.NoError(t, err)
+	staking.EndBlocker(input.Context, input.StakingKeeper)
+	qgb.EndBlocker(input.Context, *pk)
+	input.Context = ctx.WithBlockHeight(ctx.BlockHeight() + 10)
+
+	assert.Equal(t, currentAttestationNonce+1, pk.GetLatestAttestationNonce(ctx))
+}
+
+func TestValsetChangeWhenEditingOrchAddr(t *testing.T) {
+	input, ctx := testutil.SetupFiveValChain(t)
+	pk := input.QgbKeeper
+
+	// run abci methods after chain init
+	staking.EndBlocker(input.Context, input.StakingKeeper)
+	qgb.EndBlocker(ctx, *pk)
+
+	// current attestation expectedNonce should be 1 because a valset has been emitted upon chain init.
+	currentAttestationNonce := pk.GetLatestAttestationNonce(ctx)
+	require.Equal(t, uint64(1), currentAttestationNonce)
+
+	input.Context = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+	msgServer := stakingkeeper.NewMsgServerImpl(input.StakingKeeper)
 
-	// Run the staking endblocker to ensure valset is set in state
+	newOrchAddr := sdk.AccAddress(ed25519.GenPrivKey().PubKey().Address())
+	editMsg := stakingtypes.NewMsgEditValidator(
+		testutil.ValAddrs[1],
+		stakingtypes.Description{},
+		nil,
+		nil,
+		&newOrchAddr,
+		nil,
+	)
+	_, err := msgServer.EditValidator(input.Context, editMsg)
+	require.NoError(t, err)
 	staking.EndBlocker(input.Context, input.StakingKeeper)
 	qgb.EndBlocker(input.Context, *pk)
+	input.Context = ctx.WithBlockHeight(ctx.BlockHeight() + 10)
 
-	assert.NotEqual(t, currentValsetNonce, pk.GetLatestAttestationNonce(ctx))
+	// not incrementing the nonce because a change in orch address shouldn't
```

Comment:
I was confused by the test name: `TestValsetChangeWhenEditingOrchAddr` which led me to think that this test expects the validator set to change. Proposal to rename the test name to the function under test. 

Also, it isn't immediately clear to me how the attestation nonce relates to the validator set.  Maybe this can be described via a doc comment? Something like the following if it's accurate:

```go
// Assert that the attestation nonce hasn't changed because a change in the orchestrator address would manifest in a higher attestation nonce.
```

Reply:
d2253be

Reply:
let me know if it is better this way